packages/flame/lib/src/experimental/geometry/shapes/polygon.dart

```diff
@@ -0,0 +1,219 @@
+import 'dart:ui';
+
+import 'package:vector_math/vector_math_64.dart';
+
+import '../../../game/transform2d.dart';
+import 'shape.dart';
+
+/// An arbitrary polygon with 3 or more vertices.
+///
+/// The vertices of the polygon are in the counter-clockwise order, so that the
+/// polygon's interior is in your left-hand side as you traverse the polygon's
+/// vertices.
+///
+/// A polygon can be either convex or not, the [containsPoint] method will work
+/// in both cases, however, the method used with convex polygon is faster.
+class Polygon extends Shape {
+  /// Constructs the polygon from the given list of vertices.
+  ///
+  /// If the list is not in the counter-clockwise order, then it will be
+  /// reversed in-place.
+  ///
+  /// If the [convex] flag is provided, then it serves as a hint about whether
+  /// the polygon is convex or not. With this flag the user promises that the
+  /// vertices are already in the correct CCW order.
+  Polygon(this._vertices, {bool? convex})
+      : assert(_vertices.length >= 3, 'At least 3 vertices are required') {
+    _initializeEdges();
+    if (convex == null) {
+      _checkOrientation();
+    } else {
+      _convex = convex;
+    }
+  }
+
+  /// The vertices (corners) of the polygon.
+  ///
+  /// The user should treat this list as read-only and not attempt to modify
+  /// either the list itself or individual points.
+  List<Vector2> get vertices => _vertices;
+  final List<Vector2> _vertices;
+
+  /// The edges (sides) of the polygon.
+  ///
+  /// Each i-th edge is equal to the vector difference between the i-th vertex
+  /// and the preceding vertex. The number of edges is always equal to the
+  /// number of vertices.
+  List<Vector2> get edges => _edges;
+  late List<Vector2> _edges;
+  void _initializeEdges() {
+    var previousVertex = _vertices.last;
+    _edges = _vertices.map((Vector2 vertex) {
+      final edge = vertex - previousVertex;
+      previousVertex = vertex;
+      return edge;
+    }).toList(growable: false);
+  }
+
+  /// Number of vertices/edges in the polygon.
+  int get n => _vertices.length;
+
+  void _checkOrientation() {
+    var nInteriorAngles = 0;
+    var nExteriorAngles = 0;
+    var previousEdge = _edges.last;
+    _edges.forEach((edge) {
+      final crossProduct = edge.cross(previousEdge);
+      previousEdge = edge;
+      // A straight angle counts as both internal and external
+      if (crossProduct >= 0) {
+        nInteriorAngles++;
+      }
+      if (crossProduct <= 0) {
+        nExteriorAngles++;
+      }
+    });
+    if (nInteriorAngles < nExteriorAngles) {
+      _reverseVertices();
+      _initializeEdges();
+      nInteriorAngles = nExteriorAngles;
+    }
+    _convex = nInteriorAngles == n;
```

Comment:
Is the "orientation" really correct term for determining whether a shape is convex or not?

Reply:
The primary purpose of this method is to verify that the vertices are listed in the counter-clockwise order, and if not reverse the list in-place (which puts it into the CCW order). This is needed because otherwise the `containsPoint()` method won't work correctly. As a by-product, the method also detects whether the shape is convex or not.

Reply:
Hmm, I see the usefulness of setting `_convex` here, but it is a side effect of what you'd expect this method to do.

Reply:
Not sure what you're suggesting? Add a comment clarifying the purpose of the method?

Reply:
I don't have any suggestion yet, maybe tomorrow. I just don't like methods with side effects.

Reply:
Didn't have time to look properly at this today either, but I think maybe `check` is the wrong wording since it actually updates the vertices, so maybe a name that reflects that?

Reply:
Ok, I renamed it into `_ensureProperOrientation()`.

Reply:
Much better!